README
Network UPS Tools Overview
===========================

Description
-----------

Network UPS Tools is a collection of programs which provide a common
interface for monitoring and administering UPS, PDU and SCD hardware.
It uses a layered approach to connect all of the parts.

Drivers are provided for a wide assortment of equipment.  They
understand the specific language of each device and map it back to a
compatibility layer.  This means both an expensive high end UPS, a simple
"power strip" PDU, or any other power device can be handled transparently with
a uniform management interface.

This information is cached by the network server `upsd`, which then
answers queries from the clients.  upsd contains a number of access
control features to limit the abilities of the clients.  Only authorized
hosts may monitor or control your hardware if you wish.  Since the
notion of monitoring over the network is built into the software, you
can hang many systems off one large UPS, and they will all shut down
together. You can also use NUT to power on, off or cycle your data center
nodes, individually or globally through PDU outlets.

Clients such as `upsmon` check on the status of the hardware and do things
when necessary.  The most important task is shutting down the operating
system cleanly before the UPS runs out of power.  Other programs are
also provided to log information regularly, monitor status through your
web browser, and more.


Installing
----------

If you are installing these programs for the first time, go read the
<<_installation_instructions,installation instructions>>
to find out how to do that.  This document contains more information on what all
of this stuff does.


Upgrading
---------

When upgrading from an older version, always check the
<<Upgrading_notes,upgrading notes>> to see what may have
changed.  Compatibility issues and other changes will be listed there to ease
the process.


Configuring and using
---------------------

Once NUT is installed, refer to the
<<Configuration_notes,configuration notes>> for directions.


Documentation
-------------

This is just an overview of the software.  You should read the man pages,
included example configuration files, and auxiliary documentation for the parts
that you intend to use.


Network Information
-------------------

These programs are designed to share information over the network.  In
the examples below, `localhost` is used as the hostname.  This can also
be an IP address or a fully qualified domain name.  You can specify a
port number if your upsd process runs on another port.

In the case of the program `upsc`, to view the variables on the UPS called
sparky on the `upsd` server running on the local machine, you'd do this:

	/usr/local/ups/bin/upsc sparky@localhost

The default port number is 3493.  You can change this with
"configure --with-port" at compile-time.  To make a client talk to upsd
on a specific port, add it after the hostname with a colon, like this:

	/usr/local/ups/bin/upsc sparky@localhost:1234

This is handy when you have a mixed environment and some of the systems
are on different ports.

The general form for UPS identifiers is this:

	<upsname>[@<hostname>[:<port>]]

Keep this in mind when viewing the examples below.


Manifest
--------

This package is broken down into several categories:

- *drivers*	- These programs talk directly to your UPS hardware.
- *server*	- upsd serves data from the drivers to the network.
- *clients*	- They talk to upsd and do things with the status data.
- *cgi-bin*	- Special class of clients that you can use with your web server.
- *scripts*	- Contains various scripts, like the Perl and Python binding,
integration bits and applications. 

Drivers
-------

These programs provide support for specific UPS models.  They understand
the protocols and port specifications which define status information
and convert it to a form that upsd can understand.

To configure drivers, edit ups.conf.  For this example, we'll have a UPS
called "sparky" that uses the apcsmart driver and is connected to
`/dev/ttyS1`.  That's the second serial port on most Linux-based systems.
The entry in `ups.conf` looks like this:

	[sparky]
		driver = apcsmart
		port = /dev/ttyS1

To start and stop drivers, use upsdrvctl.  By default, it will start or
stop every UPS in the config file:

	/usr/local/ups/bin/upsdrvctl start
	/usr/local/ups/bin/upsdrvctl stop

However, you can also just start or stop one by adding its name:

	/usr/local/ups/bin/upsdrvctl start sparky
	/usr/local/ups/bin/upsdrvctl stop sparky

To find the driver name for your device, refer to the section below
called "HARDWARE SUPPORT TABLE".

Extra Settings
~~~~~~~~~~~~~~

Some drivers may require additional settings to properly communicate
with your hardware.  If it doesn't detect your UPS by default, check the
driver's man page or help (-h) to see which options are available.

For example, the usbhid-ups driver allows you to use USB serial numbers to
distingish between units via the "serial" configuration option.  To use this
feature, just add another line to your ups.conf section for that UPS:

	[sparky]
		driver = usbhid-ups
		port = auto
		serial = 1234567890

Hardware Compatibility List
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The <<HCL,Hardware Compatibility List>> is available in the source directory
('nut-X.Y.Z/data/driver.list'), and is generally distributed with packages.
For example, it is available on Debian systems as:

	/usr/share/nut/driver.list

This table is also available link:http://www.networkupstools.org/stable-hcl.html[online].


If your driver has vanished, see the link:FAQ.html[FAQ] and
<<Upgrading_notes,Upgrading notes>>.

Generic Device Drivers
~~~~~~~~~~~~~~~~~~~~~~

NUT provides several generic drivers that support a variety of very similar models.

- The `genericups` driver supports many serial models that use the same basic
principle to communicate with the computer.  This is known as "contact
closure", and basically involves raising or lowering signals to indicate
power status.
+
This type of UPS tends to be cheaper, and only provides the very simplest
data about power and battery status.  Advanced features like battery
charge readings and such require a "smart" UPS and a driver which
supports it.
+
See the linkman:genericups[8] man page for more information.

- The `usbhid-ups` driver attempts to communicate with USB HID Power Device
Class (PDC) UPSes. These units generally implement the same basic protocol,
with minor variations in the exact set of supported attributes. This driver
also applies several correction factors when the UPS firmware reports values
with incorrect scale factors.
+
See the linkman:usbhid-ups[8] man page for more information.

- The `blazer_ser` and `blazer_usb` drivers supports the Megatec / Q1
protocol that is used in many brands (Blazer, Energy Sistem, Fenton
Technologies, Mustek and many others).
+
See the linkman:blazer[8] man page for more information.

- The `snmp-ups` driver handles various SNMP enabled devices, from many
different manufacturers. In SNMP terms, `snmp-ups` is a manager, that
monitors SNMP agents.
+
See the linkman:snmp-ups[8] man page for more information.

- The `powerman-pdu` is a bridge to the PowerMan daemon, thus handling all
PowerMan supported devices. The PowerMan project supports several serial
and networked PDU, along with Blade and IPMI enabled servers.
+
See the linkman:powerman-pdu[8] man page for more
information.

- The `apcupsd-ups` driver is a bridge to the Apcupsd daemon, thus handling
all Apcupsd supported devices. The Apcupsd project supports many serial,
USB and networked APC UPS.
+
See the linkman:apcupsd-ups[8] man page for more information.

UPS Shutdowns
~~~~~~~~~~~~~

upsdrvctl can also shut down (power down) all of your UPS hardware.

WARNING: if you play around with this command, expect your filesystems
to die.  Don't power off your computers unless they're ready for it:

	/usr/local/ups/bin/upsdrvctl shutdown
	/usr/local/ups/bin/upsdrvctl shutdown sparky

You should read the <<UPS_shutdown,Configuring automatic UPS shutdowns>>
chapter to learn more about when to use this feature.  If called at the wrong
time, you may cause data loss by turning off a system with a filesystem
mounted read-write.

Power distribution unit management
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

NUT also provides an advanced support for power distribution units.

You should read the <<Outlets_PDU_notes,Configuring automatic UPS shutdowns>>
chapter to learn more about when to use this feature. 

Network Server
--------------

`upsd` is responsible for passing data from the drivers to the client
programs via the network.  It should be run immediately after `upsdrvctl`
in your system's startup scripts.

`upsd` should be kept running whenever possible, as it is the only source
of status information for the monitoring clients like `upsmon`.


Monitoring client
-----------------

`upsmon` provides the essential feature that you expect to find in UPS
monitoring software: safe shutdowns when the power fails.

In the layered scheme of NUT software, it is a client.  It has this
separate section in the documentation since it is so important.

You configure it by telling it about UPSes that you want to monitor in
upsmon.conf.  Each UPS can be defined as one of two possible types:

Master
~~~~~~

This UPS supplies power to the system running `upsmon`, and this system is also
responsible for shutting it down when the battery is depleted.  This occurs
after any slave systems have disconnected safely.

If your UPS is plugged directly into a system's serial port, the `upsmon`
process on that system should define that UPS as a master.

For a typical home user, there's one computer connected to one UPS.
That means you run a driver, `upsd`, and `upsmon` in master mode.

Slave
~~~~~

This UPS may supply power to the system running `upsmon`, but this system can't
shut it down directly.

Use this mode when you run multiple computers on the same UPS.  Obviously, only
one can be connected to the serial port on the UPS, and that system is the
master.  Everything else is a slave.

For a typical home user, there's one computer connected to one UPS.
That means you run a driver, upsd, and upsmon in master mode.

Additional Information
~~~~~~~~~~~~~~~~~~~~~~

More information on configuring upsmon can be found in these places:

- The linkman:upsmon[8] man page
- <<BigServers,Typical setups for big servers>>
- <<UPS_shutdown,Configuring automatic UPS shutdowns>> chapter
- The stock `upsmon.conf` that comes with the package


Clients
-------

Clients talk to upsd over the network and do useful things with the data
from the drivers.  There are tools for command line access, and a few
special clients which can be run through your web server as CGI
programs.

For more details on specific programs, refer to their man pages.

upsc
~~~~

`upsc` is a simple client that will display the values of variables known
to `upsd` and your UPS drivers.  It will list every variable by default,
or just one if you specify an additional argument.  This can be useful
in shell scripts for monitoring something without writing your own
network code.

`upsc` is a quick way to find out if your driver(s) and upsd are working
together properly.  Just run `upsc <ups>` to see what's going on, i.e.:

	morbo:~$ upsc sparky@localhost
	ambient.humidity: 035.6
	ambient.humidity.alarm.maximum: NO,NO
	ambient.humidity.alarm.minimum: NO,NO
	ambient.temperature: 25.14
	...

If you are interested in writing a simple client that monitors `upsd`,
the source code for `upsc` is a good way to learn about using the
upsclient functions.

See the linkman:upsc[8] man page and
<<nut-names,NUT command and variable naming scheme>> for more information.

upslog
~~~~~~

`upslog` will write status information from `upsd` to a file at set
intervals.  You can use this to generate graphs or reports with other
programs such as `gnuplot`.

upsrw
~~~~~

`upsrw` allows you to display and change the read/write variables in your
UPS hardware.  Not all devices or drivers implement this, so this may
not have any effect on your system.

A driver that supports read/write variables will give results like this:

	$ upsrw sparky@localhost

	( many skipped )

	[ups.test.interval]
	Interval between self tests
	Type: ENUM
	Option: "1209600"
	Option: "604800" SELECTED
	Option: "0"

	( more skipped )

On the other hand, one that doesn't support them won't print anything:

	$ upsrw fenton@gearbox

	( nothing )

`upsrw` requires administrator powers to change settings in the hardware.
Refer to linkman:upsd.users[5] for information on defining
users in `upsd`.

upscmd
~~~~~~

Some UPS hardware and drivers support the notion of an instant command -
a feature such as starting a battery test, or powering off the load.
You can use upscmd to list or invoke instant commands if your
hardware/drivers support them.

Use the -l command to list them, like this:

	$ upscmd -l sparky@localhost
	Instant commands supported on UPS [sparky@localhost]:

	load.on - Turn on the load immediately
	test.panel.start - Start testing the UPS panel
	calibrate.start - Start run time calibration
	calibrate.stop - Stop run time calibration
	...

`upscmd` requires administrator powers to start instant commands.
To define users and passwords in `upsd`, see
linkman:upsd.users[5].


CGI Programs
------------

The CGI programs are clients that run through your web server.  They
allow you to see UPS status and perform certain administrative commands
from any web browser.  Javascript and cookies are not required.

These programs are not installed or compiled by default.  To compile
and install them, first run `configure --with-cgi`, then do `make` and
`make install`.  If you receive errors about "gd" during configure, go
get it and install it before continuing.

You can get the source here:

	http://www.libgd.org/

In the event that you need libpng or zlib in order to compile gd,
they can be found at these URLs:

	http://www.libpng.org/pub/png/pngcode.html

	http://www.gzip.org/zlib/


Access Restrictions
~~~~~~~~~~~~~~~~~~~

The CGI programs use hosts.conf to see if they are allowed to talk to a
host.  This keeps malicious visitors from creating queries from your web
server to random hosts on the Internet.

If you get error messages that say "Access to that host is not
authorized", you're probably missing an entry in your hosts.conf.

upsstats
~~~~~~~~

`upsstats` generates web pages from HTML templates, and plugs in status
information in the right places.  It looks like a distant relative of
APC's old Powerchute interface.  You can use it to monitor several
systems or just focus on one.

It also can generate IMG references to `upsimage`.

upsimage
~~~~~~~~

This is usually called by upsstats via IMG SRC tags to draw either the
utility or outgoing voltage, battery charge percent, or load percent.

upsset
~~~~~~

`upsset` provides several useful administration functions through a web
interface.  You can use `upsset` to kick off instant commands on your UPS
hardware like running a battery test.  You can also use it to change
variables in your UPS that accept user-specified values.

Essentially, `upsset` provides the functions of `upsrw` and `upscmd`, but
with a happy pointy-clicky interface.

`upsset` will not run until you convince it that you have secured your
system.  You *must* secure your CGI path so that random interlopers
can't run this program remotely.  See the `upsset.conf` file.  Once you
have secured the directory, you can enable this program in that
configuration file.  It is not active by default.


Version Numbering
-----------------

The version numbers work like this: if the middle number is odd, it's a
development tree, otherwise it is the stable tree.

The past stable trees were 1.0, 1.2, 1.4, 2.0, 2.2 and 2.4, with the
latest stable tree designated 2.6.  The development trees were 1.1, 1.3,
1.5, 2.1 and 2.3.  As of the 2.4 release, there is no real development
branch anymore since the code is available through a revision control
system (namely Subversion) and snapshots.

Major release jumps are mostly due to large changes to the features
list.  There have also been a number of architectural changes which
may not be noticeable to most users, but which can impact developers.


Backwards and Forwards Compatibility
------------------------------------

The old network code spans a range from about 0.41.1 when TCP support 
was introduced up to the recent 1.4 series.  It used variable names
like STATUS, UTILITY, and LOADPCT.  Many of these names go back to the
earliest prototypes of this software from 1997.  At that point there
was no way to know that so many drivers would come along and introduce 
so many new variables and commands.  The resulting mess grew out of
control over the years.

During the 1.3 development cycle, all variables and instant commands
were renamed to fit into a tree-like structure.  There are major groups,
like input, output and battery.  Members of those groups have been
arranged to make sense - input.voltage and output.voltage compliment
each other.  The old names were UTILITY and OUTVOLT.  The benefits in
this change are obvious.

The 1.4 clients can talk to either type of server, and can handle either
naming scheme.  1.4 servers have a compatibility mode where they can
answer queries for both names, even though the drivers are internally
using the new format.

When 1.4 clients talk to 1.4 or 2.0 (or more recent) servers, they will
use the new names.

Here's a table to make it easier to visualize:

[options="header"]
|=============================================
|                   4+| Server version
| *Client version*    | 1.0 | 1.2 | 1.4 | 2.0+
| 1.0                 | yes | yes | yes | no
| 1.2                 | yes | yes | yes | no
| 1.4                 | yes | yes | yes | yes
| 2.0+                | no  | no  | yes | yes
|=============================================

Version 2.0, and more recent, do not contain backwards compatibility for
the old protocol and variable/command names.  As a result, 2.0 clients can't 
talk to anything older than a 1.4 server.  If you ask a 2.0 client to 
fetch "STATUS", it will fail.  You'll have to ask for "ups.status" 
instead.

Authors of separate monitoring programs should have used the 1.4 series
to write support for the new variables and command names.  Client
software can easily support both versions as long as they like.  If upsd
returns 'ERR UNKNOWN-COMMAND' to a GET request, you need to use REQ.


Support / Help / etc.
---------------------

If you are in need of help, refer to the
<<Support_Request,Support instructions>> in the user manual.


Hacking / Development Info
--------------------------

Additional documentation can be found in:

- the linkdoc:developer-guide[Developer Guide],
- the linkdoc:packager-guide[Packager Guide].


Acknowledgements / Contributions
--------------------------------

The many people who have participated in creating and improving NUT are
listed in the user manual <<Acknowledgements,acknowledgements appendix>>.


CHANGES
Mon Nov 20 19:02:57 UTC 2006 / Arjen de Korte <arjen@de-korte.org>

 - server/upsd.c: don't wait for DUMPDONE before forking to the background (this
   obsoletes the MAXINIT parameter)

Sat Nov 18 11:16:07 UTC 2006 / Arjen de Korte <arjen@de-korte.org>

 - server/upsd.c: add MAXINIT parameter to upsd.conf to make the time upsd waits
   for the DUMPDONE to complete configurable and try to (re)connect to a UPS, before
   declaring it stale, instead of the other way around.
 - server/conf.c: add MAXINIT parameter and MAXAGE must be greater than zero
 - lib/libupsconfig.ch: add MAXINIT parameter
 - conf/upsd.conf: add MAXINIT parameter
 - man/upsd.conf.5: add MAXINIT parameter

Thu Aug 31 15:07:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - drivers/mains.c : some bugs corrected
 - lib/libupsconfig.c : now allow, when a list of string is expected for a variable, to pu a string
   this string will be considered as a list of string with only one element. this will probably prevent
   a lot of wondering from the users, being more flexible. ( and with only one element, it is perhaps more
   natural to put something like allowfrom = "localhost" than allowfrom = { "localhost" }. Moreover, the
   users won't have to know if the variable need a list or not : if they have only one value, they can alway use string,
   if they have more than one, it is obvious it is a list of string that is needed ^_^

Thu Aug 31 11:31:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - lib/libupsconfig.ch : added the function get_comments_templates
 - utils/upsconfig.c : modified to use the function get_comments_templates
 - utils/migrateconfig.c : modified to use the function get_comments_templates

Thu Aug 31 90:57:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - lib/libupsconfig.c : Arnaud defined what need to be saved in configuration file
   when in mode "pm". Modified the code accordingly.

Thu Aug 31 09:25:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - clients/upsmon.c : correction of a little bug
 - lib/libupsconfig.ch : function that returns integer now return -1 if they failed, not 0 (because 0 can be a valide value)
 - utils/migrateconfig.c : some minor modifications
 - utils/upsconfig.c : added the -g option, to get the value of a variable in the configuration file. (for instance to get the powerdownflag in shutdown script)
 - man/upsconfig.8 : added the -g option
 - server/user.c : correction of some bugs

Tue Aug 29 10:11:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - lib/libupsconfig.ch : disabled the saving in configuration
   file of the notion of rights. For more info, look just below
   the header of libupsconfig.h
 - conf/comments/conf.comm.C : removed references to the rights notion
   save of the file with those reference has been saved and will be put
   on alioth for future when it will be enable again
 - man/nut.conf.5 : same comment than for conf.comm.C

Fri Aug 25 10:54:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - man/nut.conf : added
 - man/upsd.8
      /upsmon.8     Modified for the new configuration
      /nutupsdrv.8
 - utils/migrateconfig : Modified to handle multiple accept reject occurence
 - server/conf.c : Modified to take into account reject then accept
 - server/user.c : added actions FSD by default for upsmon user (master and slave)

Thu Aug 24 14:54:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - Many files I created : cleaning to replace // comment by /* */ comments
 - drivers/main.c upsdrvctl : global variable for ups.conf are now in nut.ups.global.
 - lib/libupsconfig.c : When saving, only try to save existant configuration files 
   (not all configuration files possible XD)
 - utils : renaming migratetool into migrateconfig
 - man : added manual for migrateconfig and upsconfig

Wed Aug 9  17:06:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - utils/migratetool.c : A tool to migrate from old config format to the new one
 - utils/Makefile.in : added the rules to compile migratetool
 - lib/libupsconfig.c : little to prevent a segfault in case the RUN_AS_USER group don't exist
 - common/datatypes.c : minor changes
 - utils/upsconfig.c : minor changes

Fri Aug 4  15:44:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - client/upsmon.c : adaptated to use libupsconfig instead of parseconf
 - drivers/main.ch, upsdrvctl.c : adaptated to use libupsconfig
 - server/access.ch, conf.c, user.ch : adapted to use libupsconfig
 - Makefile.in in the client, driver and server directory. gendb in drivers directory
   Adapted to use libupsconfig instead of parseconf or upsconf
 - lib/makefile : adapted to make the libupsconfig.a library
 - lib/libupsconfig.ch : added some functions
 - common/data_types.c : corrected a omission
 - data/base_config/nut.conf : Some minor changes
 - utils/Makefile.in : adapted to use libupsconfig as a library
 - utils/upsconfig.c : a minor change
 - Makefile : Changed the order of the subdir to make libupsconfig.a before what need it

Tue Aug 1  17:06:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - utils/upsconfig.c : fully functionnal and commented version of upsconfig !!
 - lib/libupsconfig.h : fully commented
 - lib/libupsconfig.c : some minor fixes
 - include/data_type.h stack.h tree.h : added comments

Mon Jul 31 16:17:00 UTC 2006 / Jonathan Dion <dion.jonathan@gmail.com>

 - common/data_types.ch : added some conversion functions
 - nutparser.ch : some minor changes
 - tree.ch : functions do not anymore exit on error but return an error value
 - data/base_config : a new directory containing template to generate configuration files
 - lib/libupsconfig.ch : lot of little changes. functions do not anymore exit on error 
   but return an error value.
 - utils/upsconfig.c : creation of this tool.
   * utils/Makefile.in added too.
 - configure.in : added a rule to make utils/Makefile
 - Makefile.in : added the utils directory in the SUBDIR variable


drivers/sms.c
/* sms.c - driver for SMS UPS hardware

   Copyright (C) 2001  Marcio Gomes  <tecnica@microlink.com.br>

   based on fentonups.c:

   Copyright (C) 1999  Russell Kroll <rkroll@exploits.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

   2001/05/17 - Version 0.10 - Initial release
   2001/06/01 - Version 0.20 - Add Battery Informations in driver
   2001/06/04 - Version 0.30 - Updated Battery Volts range, to reflect a correct
                               percent ( % )
   2002/12/02 - Version 0.40 - Update driver to new-model, based on Fentonups 
                               driver version 0.90  
   2002/12/18 - Version 0.50 - Add Sinus Single 2 KVA in database, test with
                               new Manager III sinusoidal versions.
                               Change Detect Name, SMS do not pass real Models
                               in Megatec Info command, only the VA/KVA version
   2002/12/18 - Version 0.51 - Updated Battery Volts range, to reflect a correct
                               percent ( % )
   2002/12/27 - Version 0.60 - Add new UPS Commands SDRET, SIMPWF change BTEST1
   2002/12/28 - Version 0.70 - Add new UPS Commands SHUTDOWN,STOPSHUTD,WATCHDOG
   2003/06/11 - Version 0.71 - Converted to dstate calls and names (rkroll)


   Microlink ISP/Pop-Rio contributed with MANAGER III 1300, MANAGER III 650 UPS
   and Sinus Single 2 KVA for my tests.

   http://www.microlink.com.br and http://www.pop-rio.com.br



*/

#define DRV_VERSION "0.72"

#include "main.h"
#include "serial.h"
#include "sms.h"

#define ENDCHAR 13	/* replies end with CR */

static	int	cap_upstemp = 0;
static	float	lowvolt = 0, voltrange;
static	int	lownorm, highnorm, poll_failures = 0;


static ups_info_t sms_supported_ups[] = {
	/* Vendor, Product, ExtraInfo, Description */
	{"SMS (Brazil)", "Manager III", 0, 0},

	{0, 0, 0, 0} /* End of list */
};

static void guessmodel(const char *raw)
{
	char	mch, *mstr;

	mch = raw[17];

        printf ("0         1         2        3         \n");
        printf ("012345678901234567890123567890123456789\n");
        printf ("%s\n", raw);
	mstr = xstrdup(&raw[18]);
	mstr[10] = '\0';	/* 10 chars max, per the protocol */
       
	/* trim whitespace */
	rtrim(mstr, ' ');
     
	dstate_setinfo("ups.model", "SMS %s", mstr);   
	cap_upstemp = 1;

	free(mstr);
}

static void getbaseinfo(void)
{
	char	temp[256], model[32], *raw;
	int	modelnum, i;

	/* dummy read attempt to sync - throw it out */
	ser_send(upsfd, "I\r");
	ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);

	/* now retrieve information and parse */
	ser_send(upsfd, "I\r");
	ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);
	raw = xstrdup(temp);

	if (temp[0] != '#')
		fatalx("Bad UPS info start character [%s]", temp);

	temp[11] = 0;
	temp[27] = 0;

	/* manufacturer */
	rtrim(&temp[1], ' ');

	dstate_setinfo("ups.mfr", "%s", &temp[1]);

/*
        0         1         2        3         
        012345678901234567890123567890123456789
        #SMS LTDA         1300 VA   VER 1.0 
        #SMS LTDA        1300VA SEN VER 5.0 
        #SMS LTDA             2 KVA   VER 1.0 
*/
	/* grab full model string */

	rtrim(&temp[17], ' ');

	snprintf(model, sizeof(model), "%s", &temp[17]);
/*        printf("->%s<-\n",model);  */
	modelnum = -1;

	/* figure out official model name and voltage info from table */
	for (i = 0; modeltab[i].mtext != NULL; i++) {
		if (!strcmp(modeltab[i].mtext, model)) {
			modelnum = i;
			lowvolt = modeltab[i].lowvolt;
			voltrange = modeltab[i].voltrange;
			cap_upstemp = modeltab[i].has_temp;
			break;
		}
	}

	/* table lookup fails -> guess */
	if (modelnum == -1)
		guessmodel (raw);
	else {
		dstate_setinfo("ups.model", "%s", modeltab[modelnum].desc);

		dstate_setinfo("input.transfer.low", "%i", 
			modeltab[modelnum].lowxfer);
		dstate_setinfo("input.transfer.high", "%i", 
			modeltab[modelnum].highxfer);

		lownorm = modeltab[modelnum].lownorm;
		highnorm = modeltab[modelnum].highnorm;
	}

	/* now add instant command support info */
	dstate_addcmd("test.battery.start");
	dstate_addcmd("test.battery.stop");
	dstate_addcmd("shutdown.return");	/* was CMD_SDRET */
	dstate_addcmd("test.failure.start");
	dstate_addcmd("shutdown.stayoff");	/* was CMD_SHUTDOWN */
	dstate_addcmd("shutdown.stop");
	dstate_addcmd("reset.watchdog");

	printf("Detected %s on %s\n", dstate_getinfo("ups.model"), device_path);
	free(raw);

	/* paranoia - cancel any shutdown that might already be running */
	ser_send(upsfd, "C\r");
}

static int instcmd(const char *cmdname, const char *extra)
{
	if (!strcasecmp(cmdname, "test.battery.stop")) {
		ser_send(upsfd, "CT\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "test.battery.start")) {
		ser_send(upsfd, "TL\r");/* start battery test until bat low */
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "test.failure.start")) {
		ser_send(upsfd, "T\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.return")) {
		/* shutdown and restart */
		ser_send(upsfd, "C\r");
		ser_send(upsfd, "S.3R0003\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.stayoff")) {
		/* shutdown now (one way) */
		ser_send(upsfd, "C\r");
		ser_send(upsfd, "S.3\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "shutdown.stop")) {
		/* Cancel Shutdown  */
		ser_send(upsfd, "C\r");
		return STAT_INSTCMD_HANDLED;
	}

	if (!strcasecmp(cmdname, "reset.watchdog")) {
		/* WATCHDOG Crontab Function */
		ser_send(upsfd, "C\r");
		ser_send(upsfd, "S05R0003\r");
		return STAT_INSTCMD_HANDLED;
	}

	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}

void upsdrv_initinfo(void)
{
	getbaseinfo();

	upsh.instcmd = instcmd;
	dstate_setinfo("driver.version.internal", "%s", DRV_VERSION);
}

static void pollfail(const char *why)
{
	poll_failures++;

	/* don't spew into the syslog forever */
	if (poll_failures < 3)
		upslogx(LOG_ERR, why);

	return;
}

void upsdrv_updateinfo(void)
{
	char	temp[256], utility[16], loadpct[16], acfreq[16], battvolt[16],
		upstemp[16], pstat[16], outvolt[16];
	int	util, ret;
	double	bvoltp;

	ser_send(upsfd, "Q1\r");

	ret = ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);

	/* sanity checks for poll data */
	if (strlen(temp) < 46) {
		pollfail("Poll failed: short read from UPS");
		dstate_datastale();
		return;
	}

	if (strlen(temp) > 46) {
		pollfail("Poll failed: oversized read from UPS");
		dstate_datastale();
		return;
	}

	if (temp[0] != '(') {
		pollfail("Poll failed: invalid start character");
		dstate_datastale();
		return;
	}

	if (poll_failures > 0)
		upslogx(LOG_NOTICE, "UPS poll succeeded");

	poll_failures = 0;		

	/* (MMM.M NNN.N PPP.P QQQ RR.R S.SS TT.T  b7b6b5b4b3b2b1b0<cr>
	 *
	 * MMM.M : input voltage (utility)
	 * NNN.N : fault voltage (ignored)
	 * PPP.P : output voltage
	 */

	sscanf(temp, "%*c%s %*s %s %s %s %s %s %s", utility, outvolt, loadpct,
		acfreq, battvolt, upstemp, pstat);

	dstate_setinfo("output.voltage", "%s", outvolt);
	dstate_setinfo("input.voltage", "%s", utility);
	dstate_setinfo("battery.voltage", "%s", battvolt);

	bvoltp = ((atof(battvolt) - lowvolt) / voltrange) * 100.0;

	if (bvoltp > 100.0)
		bvoltp = 100.0;

	dstate_setinfo("battery.charge", "%02.1f", bvoltp);

	status_init();

	util = atoi(utility);

	if (pstat[0] == '0') {
		status_set("OL");		/* on line */

		/* only allow these when OL since they're bogus when OB */
		if (pstat[2] == '1') {		/* boost or trim in effect */
			if (util < lownorm)
				status_set("BOOST");

			if (util > highnorm)
				status_set("TRIM");
		}

	} else {
		status_set("OB");		/* on battery */
	}

	if (pstat[1] == '1')
		status_set("LB");		/* low battery */

	status_commit();

	if (cap_upstemp == 1)
		dstate_setinfo("ups.temperature", "%s", upstemp);

	dstate_setinfo("input.frequency", "%s", acfreq);
	dstate_setinfo("ups.load", "%s", loadpct);

	dstate_dataok();
}

/* power down the attached load immediately */
void upsdrv_shutdown(void)
{
	char	temp[256], pstat[32];

	/* basic idea: find out line status and send appropriate command */

	ser_send(upsfd, "Q1\r");
	ser_get_line(upsfd, temp, sizeof(temp), ENDCHAR, "", 3, 0);
	sscanf (temp, "%*s %*s %*s %*s %*s %*s %*s %s", pstat);

	/* on battery: send S01<cr>, ups will return by itself on utility */
	/* on line: send S01R0003<cr>, ups will cycle and return soon */

	ser_send(upsfd, "S01");

	if (pstat[0] == '0') {			/* on line */
		printf("On line, sending shutdown+return command...\n");
		ser_send(upsfd, "R0003");
	} else
		printf("On battery, sending normal shutdown command...\n");

	ser_send_char(upsfd, 13);	/* end sequence */
}

void upsdrv_help(void)
{
}

void upsdrv_makevartable(void)
{
}

void upsdrv_banner(void)
{
	printf("Network UPS Tools - SMS UPS driver %s (%s)\n", 
		DRV_VERSION, UPS_VERSION);
        printf("by Marcio Gomes at Microlink - tecnica@microlink.com.br\n\n");
}

void upsdrv_initups(void)
{
	upsfd = ser_open(device_path);
	ser_set_speed(upsfd, device_path, B2400);
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}

void upsdrv_print_ups_list(void)
{
	int i = 0;
	
	printf("List of supported UPSs\n");
	while (sms_supported_ups[i].Vendor != 0 &&  sms_supported_ups[i].Name != 0) {		
		printf("===\n");
		if (sms_supported_ups[i].Name != 0)
			printf("Name      : %s\n", sms_supported_ups[i].Name);
		if (sms_supported_ups[i].Vendor != 0)
			printf("Vendor    : %s\n", sms_supported_ups[i].Vendor);
		if (sms_supported_ups[i].ExtraInfo != 0)
			printf("ExtraInfo : %s\n", sms_supported_ups[i].ExtraInfo);
		if (sms_supported_ups[i].Desc != 0)
			printf("Desc      : %s\n", sms_supported_ups[i].Desc);
		i++;
	}
}
